satpy/tests/reader_tests/test_goes_nc.py

```diff
@@ -0,0 +1,511 @@
+            if cols:
+                tab.append([c.text for c in cols])
+        return tab
+
+    def _denoise(self, string):
+        return string.replace('\n', '').replace(' ', '')
+
+    def _float(self, string):
+        """Convert string to float
+
+        Take care of numbers in exponential format
+        """
+        string = self._denoise(string)
+        exp_match = re.match('^[-.\d]+x10-(\d)$', string)
+        if exp_match:
+            exp = int(exp_match.groups()[0])
+            fac = 10 ** -exp
+            string = string.replace('x10-{}'.format(exp), '')
+        else:
+            fac = 1
+
+        return fac * float(string)
+
+
+class GOESNCFileHandlerTest(unittest.TestCase):
+
+    longMessage = True
+
+    @mock.patch('satpy.readers.goes_nc.xr')
+    def setUp(self, xr_):
+        # Disable logging
+        logging_off()
+        logging.getLogger("urllib3").setLevel(logging.WARNING)
+
+        from satpy.readers.goes_nc import GOESNCFileHandler, CALIB_COEFS
+
+        self.coefs = CALIB_COEFS['GOES-15']
+        self.all_coefs = CALIB_COEFS
+        self.channels = sorted(self.coefs.keys())
+        self.ir_channels = sorted([ch for ch in self.channels
+                                   if not GOESNCFileHandler._is_vis(ch)])
+        self.vis_channels = sorted([ch for ch in self.channels
+                                    if GOESNCFileHandler._is_vis(ch)])
+
+        # Mock file access to return a fake dataset. Choose a medium count value
+        # (100) to avoid elements being masked due to invalid
+        # radiance/reflectance/BT
+        nrows = ncols = 300
+        self.counts = 100 * 32 * np.ones((1, nrows, ncols))  # emulate 10-bit
+        self.lon = np.zeros((nrows, ncols))  # Dummy
+        self.lat = np.repeat(np.linspace(-150, 150, nrows), ncols).reshape(
+            nrows, ncols)  # Includes invalid values to be masked
+        self.time = datetime.datetime(2018, 8, 16, 16, 7)
+
+        xr_.open_dataset.return_value = xr.Dataset(
+            {'data': xr.DataArray(data=self.counts, dims=('time', 'yc', 'xc')),
+             'lon': xr.DataArray(data=self.lon,  dims=('yc', 'xc')),
+             'lat': xr.DataArray(data=self.lat, dims=('yc', 'xc')),
+             'time': xr.DataArray(data=np.array([self.time],
+                                                dtype='datetime64[ms]'),
+                                  dims=('time',)),
+             'bands': xr.DataArray(data=np.array([1]))},
+            attrs={'Satellite Sensor': 'G-15'})
+
+        # Instantiate reader using the mocked open_dataset() method
+        self.reader = GOESNCFileHandler(filename='dummy', filename_info={},
+                                        filetype_info={})
+
+    def test_coefs(self):
+        """Test calibration coefficients against NOAA reference"""
+        try:
+            cc = GOESCoefficientCollector()
+        except ImportError:
+            self.skipTest("This test requires bs4")
```

Comment:
I'm really not a fan of unit tests downloading files to check things like this. I run unit tests *a lot* as I change things in satpy and this seems wasteful. That said, I can't think of a better way to handle this so I guess it stays. @mraspaud @pnuu @adybbroe FYI see GOESCoefficientCollector class further up the file. Adds bs4 and html5lib as test dependencies.

Reply:
I could add the html pages as static test data? I guess they won't change in the future.

Reply:
But then we are making sure that the coefficients in the python code are up to date with static/unchanging/local HTML tables? At that point why are we checking them at all, right? Or am I missing something? The HTML isn't checked in the regular reader code, right?

I'm not familiar enough with the older GOES satellites so you'll have to let me know if the coefficients really will *never* change. I have a weird feeling about this because I don't like downloading things every time a test is run, but I also don't want to lose the code you wrote to check this. 😉

Reply:
No, the HTML is not checked in the regular reader code. I just wanted to make sure that the coefficients in the python code match the HTML page :)

The collector is not as dynamic as you may think. When the coefficients will be updated they probably get a new revision and a new table and hence will not be recognized by the collector. Some kind of manual intervention would be required anyway. So a static file would be sufficient.

Reply:
Did you use the code in this test module to generate the python code/dict that is in the reader? Maybe we should add it to the reader code and test it with static HTML files as a "feature" that isn't actively used. Then when the tables get updated we assume that someone in the pytroll community will know this and will start the process to have them updated (meaning no downloads during tests).

Another thought with downloading this during unit tests is a new user who discovers the failure first and then gets overwhelmed trying to figure out why their tests are failing.